File: apps/ios-expo/src/lib/interstitialPolicy.ts
// Pure interstitial frequency policy. NO runtime/native imports so it is
// unit-testable in isolation. The imperative ad lifecycle lives in
// ./interstitial.ts and consumes shouldShow().

export const SHOW_EVERY = 3; // show on every Nth qualifying event
export const MIN_INTERVAL_MS = 180_000; // 180s floor between interstitials
export const SESSION_CAP = 3; // max interstitials per app session

export interface FreqState {
  count: number; // qualifying events seen this session
  shownThisSession: number; // interstitials shown this session
  lastShownTs: number; // epoch ms of last shown (0 if never), persisted
  adLoaded: boolean; // is an interstitial preloaded and ready
}

export function shouldShow(s: FreqState, now: number): boolean {
  return (
    s.adLoaded &&
    s.count > 0 &&
    s.count % SHOW_EVERY === 0 &&
    s.shownThisSession < SESSION_CAP &&
    now - s.lastShownTs >= MIN_INTERVAL_MS
  );
}

File: apps/ios/RateRadar/Services/InterstitialPolicy.swift
import Foundation

/// Pure interstitial frequency policy — 1:1 port of
/// apps/ios-expo/src/lib/interstitialPolicy.ts (unit-tested there and here).
///
/// Show on every 3rd qualifying event, max 3 per session, at least 180s apart
/// (last-shown timestamp persisted across launches).
struct InterstitialPolicy {
    static let showEvery = 3
    static let sessionCap = 3
    static let minIntervalMs: Double = 180_000

    struct State {
        var count: Int = 0
        var adLoaded: Bool = false
        var shownThisSession: Int = 0
        /// Epoch milliseconds of the last shown interstitial (0 = never).
        var lastShownTs: Double = 0
    }

    static func shouldShow(_ s: State, nowMs: Double) -> Bool {
        s.adLoaded
            && s.count > 0
            && s.count % showEvery == 0
            && s.shownThisSession < sessionCap
            && nowMs - s.lastShownTs >= minIntervalMs
    }
}
